src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs

```diff
@@ -302,7 +345,29 @@ private Expression NullCheck(Expression input)
             }
         }
 
-        internal class ConstantExpressionReplacementVisitor : ExpressionVisitor
+        private class ObjectParameterExpressionReplacementVisitor : ExpressionVisitor
+        {
+            private readonly object _parameters;
+
+            public ObjectParameterExpressionReplacementVisitor(object parameters) => _parameters = parameters;
+
+            protected override Expression VisitMember(MemberExpression node)
+            {
+                if (!node.Member.DeclaringType.Has<CompilerGeneratedAttribute>())
+                {
+                    return base.VisitMember(node);
+                }
+                var parameterName = node.Member.Name;
+
+                var matchingMember = _parameters.GetType().GetDeclaredProperty(parameterName);
```

Comment:
There is some reflection here on very call, but we already have that with CompilerGeneratedAttribute above.